src/vmm/src/vstate/vcpu/mod.rs

```diff
@@ -315,6 +361,46 @@ impl Vcpu {
                     )))
                     .expect("vcpu channel unexpectedly closed");
             }
+            #[cfg(feature = "debug")]
+            Ok(VcpuEvent::GetRegisters) => {
+                self.response_sender
+                    .send(VcpuResponse::NotAllowed(String::from(
+                        "get registers is unavailable while running",
+                    )))
+                    .expect("vcpu channel unexpectedly closed");
+            }
+            #[cfg(feature = "debug")]
+            Ok(VcpuEvent::SetRegisters(_)) => {
+                self.response_sender
+                    .send(VcpuResponse::NotAllowed(String::from(
+                        "dump is unavailable while running",
+                    )))
+                    .expect("vcpu channel unexpectedly closed");
+            }
+            #[cfg(feature = "debug")]
+            Ok(VcpuEvent::SetKvmDebug(_, _)) => {
+                self.response_sender
+                    .send(VcpuResponse::NotAllowed(String::from(
+                        "kvm debug is unavailable while running",
+                    )))
+                    .expect("vcpu channel unexpectedly closed");
+            }
+            #[cfg(feature = "debug")]
+            Ok(VcpuEvent::InjectKvmBP(_, _)) => {
+                self.response_sender
+                    .send(VcpuResponse::NotAllowed(String::from(
+                        "Injecting bp is unavailable while running",
+                    )))
+                    .expect("vcpu channel unexpectedly closed");
+            }
+            #[cfg(feature = "debug")]
+            Ok(VcpuEvent::GvaTranslate(_)) => {
+                self.response_sender
+                    .send(VcpuResponse::NotAllowed(String::from(
+                        "gva translate is unavailable while running",
+                    )))
+                    .expect("vcpu channel unexpectedly closed");
+            }
```

Comment:
Another comment that's more about the existing code, but: Could `Vcpu::NotAllowed` be made to contain a `VcpuEvent` instead of a string? Then all these "not allowed" cases would turn into
```suggestion
            Ok(event @ (VcpuEvent::GetRegisters | VcpuEvent::SetRegister(_) | VcpuEvent::SetKvmDebug(_, _) | VcpuEvent::InjectKvmBP(_, _) | VcpuEvent::GvaTranslate(_))) => {
                self.response_sender
                    .send(VcpuResponse::NotAllowed(event))
                    .expect("vcpu channel unexpectedly closed");
            }
```

in fact, here's the diff that does the refactor, I think we should definitely do this :pleading_face::

<details>

```diff
diff --git a/src/vmm/src/lib.rs b/src/vmm/src/lib.rs
index 6a681431f..79d6de1d1 100644
--- a/src/vmm/src/lib.rs
+++ b/src/vmm/src/lib.rs
@@ -294,8 +294,8 @@ pub enum DumpCpuConfigError {
     UnexpectedResponse,
     /// Failed to dump CPU config: {0}
     DumpCpuConfig(#[from] vcpu::VcpuError),
-    /// Operation not allowed: {0}
-    NotAllowed(String),
+    /// Operation not allowed: {0:?}
+    NotAllowed(VcpuEvent),
 }
 
 /// Contains the state and associated methods required for the Firecracker VMM.
diff --git a/src/vmm/src/persist.rs b/src/vmm/src/persist.rs
index 6410e4d6c..0b391effb 100644
--- a/src/vmm/src/persist.rs
+++ b/src/vmm/src/persist.rs
@@ -41,7 +41,7 @@ use crate::vstate::memory::{
 };
 use crate::vstate::vcpu::{VcpuSendEventError, VcpuState};
 use crate::vstate::vm::VmState;
-use crate::{mem_size_mib, vstate, EventManager, Vmm, VmmError};
+use crate::{mem_size_mib, vstate, EventManager, Vmm, VmmError, VcpuEvent};
 
 /// Holds information related to the VM that is not part of VmState.
 #[derive(Clone, Debug, Default, Deserialize, PartialEq, Eq, Serialize)]
@@ -115,8 +115,8 @@ pub enum MicrovmStateError {
     IncompatibleState(String),
     /// Provided MicroVM state is invalid.
     InvalidInput,
-    /// Operation not allowed: {0}
-    NotAllowed(String),
+    /// Operation not allowed: {0:?}
+    NotAllowed(VcpuEvent),
     /// Cannot restore devices: {0}
     RestoreDevices(DevicePersistError),
     /// Cannot restore Vcpu state: {0}
diff --git a/src/vmm/src/vstate/vcpu/mod.rs b/src/vmm/src/vstate/vcpu/mod.rs
index 366e7555c..3232412c1 100644
--- a/src/vmm/src/vstate/vcpu/mod.rs
+++ b/src/vmm/src/vstate/vcpu/mod.rs
@@ -346,59 +346,15 @@ impl Vcpu {
                     .expect("vcpu channel unexpectedly closed");
             }
             // SaveState cannot be performed on a running Vcpu.
-            Ok(VcpuEvent::SaveState) => {
-                self.response_sender
-                    .send(VcpuResponse::NotAllowed(String::from(
-                        "save/restore unavailable while running",
-                    )))
-                    .expect("vcpu channel unexpectedly closed");
-            }
-            // DumpCpuConfig cannot be performed on a running Vcpu.
-            Ok(VcpuEvent::DumpCpuConfig) => {
+            Ok(event @ (VcpuEvent::SaveState | VcpuEvent::DumpCpuConfig)) => {
                 self.response_sender
-                    .send(VcpuResponse::NotAllowed(String::from(
-                        "cpu config dump is unavailable while running",
-                    )))
+                    .send(VcpuResponse::NotAllowed(event))
                     .expect("vcpu channel unexpectedly closed");
             }
             #[cfg(feature = "debug")]
-            Ok(VcpuEvent::GetRegisters) => {
+            Ok(event @ (VcpuEvent::GetRegisters | VcpuEvent::SetRegisters(_) | VcpuEvent::SetKvmDebug(_, _) | VcpuEvent::InjectKvmBP(_, _) | VcpuEvent::GvaTranslate(_))) => {
                 self.response_sender
-                    .send(VcpuResponse::NotAllowed(String::from(
-                        "get registers is unavailable while running",
-                    )))
-                    .expect("vcpu channel unexpectedly closed");
-            }
-            #[cfg(feature = "debug")]
-            Ok(VcpuEvent::SetRegisters(_)) => {
-                self.response_sender
-                    .send(VcpuResponse::NotAllowed(String::from(
-                        "dump is unavailable while running",
-                    )))
-                    .expect("vcpu channel unexpectedly closed");
-            }
-            #[cfg(feature = "debug")]
-            Ok(VcpuEvent::SetKvmDebug(_, _)) => {
-                self.response_sender
-                    .send(VcpuResponse::NotAllowed(String::from(
-                        "kvm debug is unavailable while running",
-                    )))
-                    .expect("vcpu channel unexpectedly closed");
-            }
-            #[cfg(feature = "debug")]
-            Ok(VcpuEvent::InjectKvmBP(_, _)) => {
-                self.response_sender
-                    .send(VcpuResponse::NotAllowed(String::from(
-                        "Injecting bp is unavailable while running",
-                    )))
-                    .expect("vcpu channel unexpectedly closed");
-            }
-            #[cfg(feature = "debug")]
-            Ok(VcpuEvent::GvaTranslate(_)) => {
-                self.response_sender
-                    .send(VcpuResponse::NotAllowed(String::from(
-                        "gva translate is unavailable while running",
-                    )))
+                    .send(VcpuResponse::NotAllowed(event))
                     .expect("vcpu channel unexpectedly closed");
             }
             Ok(VcpuEvent::Finish) => return StateMachine::finish(),
@@ -752,7 +708,7 @@ pub enum VcpuResponse {
     /// Vcpu is stopped.
     Exited(FcExitCode),
     /// Requested action not allowed.
-    NotAllowed(String),
+    NotAllowed(VcpuEvent),
     /// Vcpu is paused.
     Paused,
     /// Vcpu is resumed.
@@ -778,7 +734,7 @@ impl fmt::Debug for VcpuResponse {
             Exited(code) => write!(f, "VcpuResponse::Exited({:?})", code),
             SavedState(_) => write!(f, "VcpuResponse::SavedState"),
             Error(ref err) => write!(f, "VcpuResponse::Error({:?})", err),
-            NotAllowed(ref reason) => write!(f, "VcpuResponse::NotAllowed({})", reason),
+            NotAllowed(ref reason) => write!(f, "VcpuResponse::NotAllowed({:?})", reason),
             DumpedCpuConfig(_) => write!(f, "VcpuResponse::DumpedCpuConfig"),
             #[cfg(feature = "debug")]
             KvmRegisters(_) => write!(f, "VcpuResponse::KvmRegisters"),
```

</details>

Reply:
I've left this for now as we don't touch this now with the most recent change,